### PR TITLE
Add reverse binomial with tests.

### PR DESCRIPTION
**Pull request by hernanmd**
See https://math.stackexchange.com/a/103385/205 for details

### CODE REVIEW

File: src/Math-Numerical/Integer.extension.st [hunk 1]
```diff
@@ -1,5 +1,54 @@
 Extension { #name : #Integer }
 
+{ #category : #'*Math-Numerical' }
+Integer >> findNK [
+	" Reverse binomial coefficient. Answer a <Collection> with all n and k such that n take: k = self. Elements in the answered Collection should be read as paired. Each pair represents (n,k) in the binomial coefficient formula. 
+	See https://math.stackexchange.com/a/103385/205 for details. "
+
+	| k |
+	[ self isInteger and: [ self > 1 ] ] assert.
```
Comment (hemalvarambhia) [inline src/Math-Numerical/Integer.extension.st:9]: the method is a little long. Is there a way to break it up?

File: src/Math-Tests-Numerical/IntegerTest.extension.st [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	self should: [ 1 findNK ] raise: Error.
```
Comment (hemalvarambhia) [inline src/Math-Tests-Numerical/IntegerTest.extension.st:15]: Move these to test methods as proposed by @olekscode perhaps?

File: src/Math-Numerical/Integer.extension.st [hunk 3]
```diff
@@ -1,5 +1,54 @@
 Extension { #name : #Integer }
 
+{ #category : #'*Math-Numerical' }
+Integer >> findNK [
```
Comment (hemalvarambhia) [inline src/Math-Numerical/Integer.extension.st:4]: Maybe `reverseBinomial:` given your comment  in the method block?

File: src/Math-Numerical/Integer.extension.st (hunk 1 above)
Comment (hemalvarambhia) [inline src/Math-Numerical/Integer.extension.st:9]: Given you've defined this in the Integer class, do you need to have the `isInteger` check?

File: src/Math-Numerical/Integer.extension.st [hunk 4]
```diff
@@ -1,5 +1,54 @@
...
+						stream nextPutAll: { n asInteger . k asInteger }.
+						k < (n - k)
+							ifTrue: [ stream nextPutAll: { n asInteger . (n - k) asInteger } ] ].
+	            choose := choose * (n + 1).
+	            choose := (choose / (n + 1 - k)) ceiling ] ] ]
+
+]
+
+{ #category : #'*Math-Numerical' }
+Integer >> firstOver: c [
```
Comment (hemalvarambhia) [inline src/Math-Numerical/Integer.extension.st:33]: Maybe add a link to the underlying Mathematics? This is a part of the domain I'm not clear on.

**Comment by hemalvarambhia** [comment]
As a first observation this seems to be a direct translation of the python code snippet in the stack exchange link.

File: src/Math-Numerical/Integer.extension.st (hunk 1 above)
Comment (hernanmd) [reply src/Math-Numerical/Integer.extension.st]: You're right! I removed the check. Thanks!

**Comment by hemalvarambhia** [comment]
@hernanmd as I mentioned above, the code appears to be python translated, almost literally to Smalltalk. Is this the case? 

I think `findNK` could be made more intention revealing. Could the code be refactored a little as I can't seem to follow it too easily (I'm having an off day!).

**Comment by hernanmd** [comment]
> @hernanmd as I mentioned above, the code appears to be python translated, almost literally, to Smalltalk. Is this the case?
> 
> I think `findNK` could be made more intention revealing. Could the code be refactored a little as I can't seem to follow it too easily (I'm having an off day!).

Yes, it is a translation from Python code that's why I included the link :)

I don't see how it could be refactored because Integer can easily explode.

File: src/Math-Tests-Numerical/IntegerTest.extension.st [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+	self assert: (3 firstOver: 4) equals: 2.
+	self assert: (3 firstOver: 5) equals: 2.
+	self assert: (3 firstOver: 6) equals: 2.
+
+	self assert: (4 firstOver: 2) equals: 2.
+	self assert: (4 firstOver: 3) equals: 2.
+	self assert: (4 firstOver: 4) equals: 2.
+	self assert: (4 firstOver: 5) equals: 2.
+	
+	self assert: (4 firstOver: SmallInteger maxVal) equals: 32768.
```
Comment (nicolas-cellier-aka-nice) [inline src/Math-Tests-Numerical/IntegerTest.extension.st:37]: SmallInteger maxVal will vary in 32bits vs 64bits image flavours...
Please use an explicit value

**Comment by hemalvarambhia** [comment]
Chasing up on this PR, @hernanmd . Let us know when you've addressed review comments and then we can merge

**Comment by hemalvarambhia** [comment]
@hernanmd chasing up again :)

File: src/Math-Numerical/Integer.extension.st [hunk 6]
```diff
@@ -1,5 +1,54 @@
...
```
Comment (nicolas-cellier-aka-nice) [inline src/Math-Numerical/Integer.extension.st:33]: firstOver: is not really intention revealing...

If I understand, that's solving `n^self>=c`, that is `n=(c^(1/self)) ceiling`, or in other words, that's `(c nthRoot: self) ceiling`.
Both Cuis and Squeak have `nthRoot:`. If Pharo lacks such method, then it shall be added to Pharo.

File: src/Math-Tests-Numerical/IntegerTest.extension.st (hunk 5 above)
Comment (hernanmd) [reply src/Math-Tests-Numerical/IntegerTest.extension.st]: Done, thanks!

File: src/Math-Numerical/Integer.extension.st (hunk 6 above)
Comment (hernanmd) [reply src/Math-Numerical/Integer.extension.st:33]: I quickly checked but `2 nthRoot: 3`in Squeak has different output than `2 firstOver: 3`.

**Comment by SergeStinckwich** [comment]
Move to 1.0.4 milestone

**Comment by SergeStinckwich** [comment]
Can we merge? @hemalvarambhia

**Comment by hemalvarambhia** [comment]
@SergeStinckwich I think the code quality may be improved. Plus, there has been no addressing of comments or any other activity. 

I would close it rather than merge. That way, it can reopened when the author has energy and capacity. 

What do other reviewers think?

**Comment by hernanmd** [comment]
> @SergeStinckwich I think the code quality may be improved. Plus, there has been no addressing of comments or any other activity.

Can you detail which aspect of the code could be improved?

IMHO it seems too much revision for a single method addition, I would love to know why or simply discard this addition definitely.

**Comment by SergeStinckwich** [comment]
The main concern of @hemalvarambhia is about names. Could you use better names for ``findNK`` and ``firstOver:`` ?

**Comment by SergeStinckwich** [comment]
I propose to replace ``findNK`` by ``findAllNKGivenBinomialCoeff``or ``inverseBinomialCoefficient``.
You can remove ``firstOver:``and replace directly in the code: ``nmin := k firstOver: (k factorial * self).``by ``nmin :=  (k factorial * self) nthRoot: k.``.

**Comment by hemalvarambhia** [comment]
> I propose to replace ``findNK`` by ``findAllNKGivenBinomialCoeff``or ``inverseBinomialCoefficient``.
> You can remove ``firstOver:`` (and tests) and replace directly in the code: ``nmin := k firstOver: (k factorial * self).``by ``nmin :=  (k factorial * self) nthRoot: k.``

`inverseBinomialCoeffient` feels more concise IMHO. Could you use Kent Beck's 'Refactor to composed method' to improve the maintainability. This is the only concern that I have.

**Comment by hemalvarambhia** [comment]
@hernanmd This is only my opinion and it is important that others weigh in. If more disagree with my view, ignore the following:

my concern is the structure of `findNK`. It feels very nested, with `while` and nested `ifTrue` messages so I sense there is a risk to maintainability and understandability, a key trade off for me. The method is also long. One option to improve things might be to appeal to 'refactor to composed method' (cf Kent Beck's 'Smalltalk Best Practice Pattern')

**Comment by SergeStinckwich** [comment]
Thank you Hernan.

**Review by SergeStinckwich — APPROVED** [review]

**Pull request merged**